### PR TITLE
Picard 2.0

### PR DESCRIPTION
**Pull request by Lctrs**
Currently building and running locally.

When launching the app from CLI, I got a strange error message. Don't know how to tackle it. See the screenshot below.

capture d ecran de 2018-03-27 11-18-18

_Le schéma « org.gnome.desktop.interface » n'existe pas_ translates to _Schema « org.gnome.desktop.interface » doesn't exist_

### CODE REVIEW

**Comment by bochecha** [comment]
Argh, I had actually started some time ago, but never pushed the branch.

I also had the error about gsettings, it is due to something awful Picard does to find the icon theme. We can ignore it entirely since icon themes aren't supported in Flatpak (yet?)

I'll have a look at your branch soon, and compare it with mine.

**Comment by Lctrs** [comment]
@bochecha is there any reason that libdiscid have never been included into the flatpak build ?

**Comment by bochecha** [comment]
> @bochecha is there any reason that libdiscid have never been included into the flatpak build ?

No reason, except the fact that I have no idea why it would be needed. Do you? I'm happy to add it even to the current Picard 1.x if it's useful.

**Comment by Lctrs** [comment]
It's used by Picard to calculate the discId of a CD for matching an entry in MusicBrainz to the CD present in the CD-ROM drive (so it will need access to it too). See [the documentation](https://musicbrainz.org/doc/libdiscid).

**Comment by Lctrs** [comment]
Some people [complained about it](https://community.metabrainz.org/t/discid-missing-on-my-ubuntu-16-04/351056) actually.

**Comment by bochecha** [comment]
> It's used by Picard to calculate the discId of a CD for matching an entry in MusicBrainz to the CD present in the CD-ROM drive

Ah, that's why I never realized it was missing: I don't have a CD/DVD player in my laptop. ^_^"

> Some people complained about it actually.

It would have been better if they had come and report a bug here. :disappointed: 

Anyway, let's add it to the current 1.x Flatpak, that's orthogonal to moving to 2.0.

**Comment by bochecha** [comment]
So, looking into this, there are **way too many** changes bundled all in one commit for me to be able to review them. :disappointed:

**Comment by Lctrs** [comment]
> In addition to that, one thing I saw is you bundle Qt5. Why not use org.kde.Sdk, which contains it already?

I'm not familiar with PyQt, but I understood that they generate bindings for Qt from source code. So if they update Qt5 version in `org.kde.Platform`, the bindings might become incompatible and it can potentially break the app.

**Comment by Lctrs** [comment]
@bochecha rebased.

File: org.musicbrainz.Picard.json [hunk 1]
```diff
@@ -1,56 +1,152 @@
...
         "--share=ipc",
         "--socket=x11",
+        /* Wayland access */
+        "--socket=wayland",
+        /* OpenGL access */
+        "--device=dri",
+        /* Needs to talk to the network: */
         "--share=network",
+        /* GConf */
+        "--talk-name=org.gnome.GConf",
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: GConf doesn't seem to be used any more in Picard 2.0, so we can remove this permission.

File: org.musicbrainz.Picard.json [hunk 2]
```diff
@@ -1,56 +1,152 @@
...
-    "id": "org.musicbrainz.Picard",
+    "app-id": "org.musicbrainz.Picard",
     "runtime": "org.freedesktop.Platform",
     "runtime-version": "1.6",
     "sdk": "org.freedesktop.Sdk",
     "command": "picard",
     "rename-desktop-file": "picard.desktop",
     "rename-icon": "picard",
     "finish-args": [
+        /* X11 + XShm access */
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: Why add these comments? Those options are well-known, used by almost all apps, there is nothing special to them that requires documenting… (plus, comments are theoretically invalid in JSON, so I'd rather we avoid them as much as possible)

File: org.musicbrainz.Picard.json [hunk 3]
```diff
@@ -1,56 +1,152 @@
...
         "--share=ipc",
         "--socket=x11",
+        /* Wayland access */
+        "--socket=wayland",
+        /* OpenGL access */
+        "--device=dri",
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: Is OpenGL really needed? I don't really see why Picard would use it. :confused:

File: org.musicbrainz.Picard.json [hunk 4]
```diff
@@ -1,56 +1,152 @@
...
                 },
                 {
                     "type": "file",
                     "path": "pyqt-configure",
                     "dest-filename": "configure"
                 }
             ],
+            "modules": [
+                {
+                    "name": "sip",
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: Why did you move sip after PyQt?

If there isn't a strong reason to have it after, I'd rather you didn't, to keep the diff minimal and easier to review. (and if there is a good reason for it, then please move it in its own commit)

File: org.musicbrainz.Picard.json [hunk 5]
```diff
@@ -66,35 +162,26 @@
...
                 "--enable-demuxer=flac",
                 "--enable-demuxer=mp3",
                 "--enable-demuxer=ogg",
-
                 "--disable-parsers",
                 "--enable-parser=flac",
                 "--enable-parser=vorbis",
-
                 "--disable-encoders",
-
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: I had added those empty lines on purpose, to visually separate the various "sections" of ffmpeg options.

If they are causing problems and need to be removed, I'd like that to happen in its own commit with an explanation. Otherwise, let's keep them please. :slightly_smiling_face:

File: org.musicbrainz.Picard.json [hunk 6]
```diff
@@ -180,11 +248,31 @@
...
+                    "type": "archive",
+                    "url": "https://github.com/acoustid/chromaprint/releases/download/v1.4.3/chromaprint-1.4.3.tar.gz",
+                    "sha256": "ea18608b76fb88e0203b7d3e1833fb125ce9bb61efe22c6e169a50c52c457f82"
+                }
+            ],
+            "cleanup": [
+                "/include",
+                "/lib/pkgconfig"
+            ]
+        },
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: Does Chromaprint really need to be moved? (as usual, if it does, please move it in its own commit, otherwise please keep it where it is)

File: org.musicbrainz.Picard.json [hunk 7]
```diff
@@ -140,7 +207,8 @@
             "sources": [
                 {
                     "type": "archive",
-                    "url": "http://ftp.musicbrainz.org/pub/musicbrainz/libdiscid/libdiscid-0.6.2.tar.gz", /* No HTTPS */
+                    "url": "http://ftp.musicbrainz.org/pub/musicbrainz/libdiscid/libdiscid-0.6.2.tar.gz",
+                    /* No HTTPS */
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: Why this change?

File: org.musicbrainz.Picard.json [hunk 8]
```diff
@@ -153,7 +221,7 @@
             "name": "python-pkgconfig",
             "buildsystem": "simple",
             "build-commands": [
-                "python2 setup.py install --prefix=/app"
+                "python3 setup.py install --prefix=/app"
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: The Python packaging ecosystem is slowly but steadily moving towards using Pip for everything, with Pip becoming a sort of "meta" tool knowing how to handle distutils, setuptools, flit, …

As a result, I prefer using the following for Python modules: `pip3 install --no-deps --prefix=/app .`.

Unfortunately that doesn't work with Python2 (the `pip` bundled for Python2 in the various SDKs is too old), and it also doesn't work for Picard 2.0 with Python 3 (their setup.py file is so convoluted, it breaks some assumptions from Pip… this might eventually get fixed in the future)

I'd much prefer we used Pip everywhere possible, and keep the setuptools-specific instructions only when absolutely necessary (the "picard" module itself)

File: org.musicbrainz.Picard.json [hunk 9]
```diff
@@ -167,10 +235,10 @@
             "name": "python-libdiscid",
             "buildsystem": "simple",
             "build-commands": [
-                "python2 setup.py install --prefix=/app"
+                "python3 setup.py install --prefix=/app"
             ],
             "ensure-writable": [
-                "/lib/python2.7/site-packages/easy-install.pth"
+                "/lib/python3.5/site-packages/easy-install.pth"
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: Using Pip to install things with Python 3, this becomes unnecessary and can be removed entirely. :slightly_smiling_face:

File: org.musicbrainz.Picard.json [hunk 10]
```diff
@@ -202,13 +290,14 @@
...
-                "python2 setup.py install --prefix=/app"
+                "python3 setup.py install --prefix=/app"
             ],
             "sources": [
                 {
                     "type": "archive",
-                    "url": "http://ftp.musicbrainz.org/pub/musicbrainz/picard/picard-1.4.2.tar.gz", /* No HTTPS */
-                    "sha256": "f0adee3877bbd66c827a43b57dd2233037b99f6ebfdc1700b26cb9d400992234"
+                    "url": "https://github.com/metabrainz/picard/archive/2.0.0dev5.tar.gz",
+                    /* No HTTPS */
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: Why move the comment?

File: org.musicbrainz.Picard.json [hunk 11]
```diff
@@ -1,56 +1,152 @@
...
         "--filesystem=home",
-        "--filesystem=xdg-music",
-        "--talk-name=org.gnome.GConf"
+        "--filesystem=xdg-music"
     ],
     "modules": [
-        "shared-modules/qt4/qt4-4.8.7-minimal.json",
         {
-            "name": "sip",
+            "name": "qt5-qtbase",
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: I'm currently talking with the maintainers of the KDE Sdk, to see if we can add PyQt to it, so we don't have to bundle it in Picard.

In any case, bundling Qt5 shouldn't be necessary, as they won't break the Qt ABI in a given SDK branch… unless PyQt uses some private APIs for which Qt doesn't guarantee the ABI.

File: org.musicbrainz.Picard.json [hunk 12]
```diff
@@ -1,56 +1,152 @@
 {
-    "id": "org.musicbrainz.Picard",
+    "app-id": "org.musicbrainz.Picard",
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: Please drop this change.

"id" and "app-id" are entirely equivalent, so there is no reason to change.

In addition, I was just told the following in #flatpak:

screenshot from 2018-03-28 15-23-08

So I'd actually prefer we use the newer name.

File: flathub.json [hunk 13]
```diff
@@ -0,0 +1,3 @@
+{
+    "skip-arches": ["i386"]
+}
```
Comment (bochecha) [inline flathub.json]: Should we also skip the arm and aarch64 builds then?

File: org.musicbrainz.Picard.json (hunk 3 above)
Comment (Lctrs) [reply org.musicbrainz.Picard.json]: It's used by Qt. Without it, Picard actually launches but we got some OpenGL related warnings from Qt in the CLI.

Comment (bochecha) [reply org.musicbrainz.Picard.json]: Ah, alright, thanks.

File: org.musicbrainz.Picard.json [hunk 14]
```diff
@@ -153,11 +160,11 @@
             "name": "python-pkgconfig",
             "buildsystem": "simple",
             "build-commands": [
-                "python2 setup.py install --prefix=/app"
+                "pip3 install --no-deps --prefix=/app pkgconfig-1.3.1.tar.gz"
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: Please use `pip3 install --no-deps --prefix=/app .` as I had mentioned, keeping the `archive` source type.

Comment (Lctrs) [reply org.musicbrainz.Picard.json]: Oh OK ! I missed the final dot in your comment. ^^

File: flathub.json (hunk 13 above)
Comment (Lctrs) [reply flathub.json]: See https://community.metabrainz.org/t/flatpak-version-of-picard-2-0/370699.

File: org.musicbrainz.Picard.json [hunk 15]
```diff
@@ -1,54 +1,61 @@
...
     "modules": [
-        "shared-modules/qt4/qt4-4.8.7-minimal.json",
         {
-            "name": "sip",
+            "name": "pyqt",
+            "config-opts": [
+                "--disable-static",
+                "--enable-x11"
+            ],
+            "no-autogen": true,
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: This `no-autogen` shouldn't be needed. Did you find it was?

File: org.musicbrainz.Picard.json [hunk 16]
```diff
@@ -1,54 +1,61 @@
...
     ],
     "modules": [
-        "shared-modules/qt4/qt4-4.8.7-minimal.json",
         {
-            "name": "sip",
+            "name": "pyqt",
+            "config-opts": [
+                "--disable-static",
+                "--enable-x11"
+            ],
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: I recently removed these config-opts from the Picard 1 manifest, they have never been useful (the configure script ignores them entirely).

I guess you brought them back when you rebased. Can you remove them please?

File: flathub.json (hunk 13 above)
Comment (bochecha) [reply flathub.json]: Thanks for asking them!

(I would have just gone and ask on IRC, but this is better as it will give us a URL we can point to)

File: org.musicbrainz.Picard.json [hunk 17]
```diff
@@ -1,54 +1,61 @@
...
+                        {
+                            "type": "file",
+                            "path": "sip-configure",
+                            "dest-filename": "configure"
+                        }
+                    ],
+                    "cleanup": [
+                        "/bin",
+                        "/include"
+                    ]
```
Comment (bochecha) [inline org.musicbrainz.Picard.json]: I'd rather you kept this as it was, ordered before PyQt.

I'm not a big fan of nested modules in flatpak manifests, (that doesn't work as soon as multiple modules require the same dependency) and it makes it hard to review whether you changed anything for the sip module.

Please put it back before PyQt5, let's keep this flat.

File: org.musicbrainz.Picard.json (hunk 15 above)
Comment (Lctrs) [reply org.musicbrainz.Picard.json]: I really don't know what it does. I copy pasted this from another repo. ^^'

File: pyqt-configure [hunk 18]
```diff
@@ -1,18 +1,25 @@
...
     --enable=QtCore \
     --enable=QtGui \
     --enable=QtNetwork \
+    --enable=QtWidgets \
     --no-designer-plugin \
     --no-docstrings \
+    --no-qml-plugin \
     --no-qsci-api \
+    --sysroot=/app \
+    -w \
```
Comment (bochecha) [inline pyqt-configure]: This is the same as `--verbose`, so don't add both.

The build is already quite verbose though, so to be honest I'd rather remove them both. (but if we do keep one, let's keep `--verbose` only)

File: pyqt-configure [hunk 19]
```diff
@@ -1,18 +1,25 @@
...
     --confirm-license \
     --enable=QtCore \
     --enable=QtGui \
     --enable=QtNetwork \
+    --enable=QtWidgets \
     --no-designer-plugin \
     --no-docstrings \
+    --no-qml-plugin \
     --no-qsci-api \
+    --sysroot=/app \
```
Comment (bochecha) [inline pyqt-configure]: When I was testing I didn't need this option. Did you find it to be necessary for something I missed?

Comment (bochecha) [reply pyqt-configure]: Aha!

So with `--sysroot=/app` we can remove `--bindir`, `--destdir` and `--sipdir`!

How did I miss this option in the first place, it's been there since 2014. :open_mouth:

File: pyqt-configure [hunk 20]
```diff
@@ -1,18 +1,25 @@
...
+    -w \
     --bindir=/app/bin \
     --destdir=/app/lib/python${PYTHON_VERSION}/site-packages \
-    --sipdir=/app/share/sip
+    --sipdir=/app/share/sip \
+    --verbose \
+    QMAKE_CFLAGS_RELEASE="-I/usr/include/python${PYTHON_VERSION}m/" \
+    QMAKE_CXXFLAGS_RELEASE="-I/usr/include/python${PYTHON_VERSION}m/"
```
Comment (bochecha) [inline pyqt-configure]: These last 2 don't seem to be needed, I can build fine without them. Am I missing something?

**Comment by Lctrs** [comment]
@bochecha Can you trigger a build to see if it builds on all arches ? Don't know if I can do it myself.

**Comment by bochecha** [comment]
bot, build

**Comment by bochecha** [comment]
@Lctrs Here's the build: https://flathub.org/builds/#/builders/2/builds/470

Also, if you want to try out builds for all arches on your own computer (beware, it's extremely slow, but it's super useful to debug things) : https://mathieu.daitauha.fr/blog/2017/08/09/building-apps-for-other-arches-with-flatpak/

**Comment by Lctrs** [comment]
Builds are green. o/

File: org.musicbrainz.Picard.json (hunk 11 above)
Comment (Lctrs) [reply org.musicbrainz.Picard.json]: We can maybe create an extension ourselves (like https://github.com/flathub/org.freedesktop.Sdk.Extension.rust-stable for example) ?

**Comment by Lctrs** [comment]
Since my first successful build of Picard 2.0, saving a file functionality is broken with this error : 

```
E: 13:45:36,622 /app/lib/python3.5/site-packages/picard/util/thread.run:51: Traceback (most recent call last):
  File "/app/lib/python3.5/site-packages/mutagen/_util.py", line 169, in wrapper
    return func(*args, **kwargs)
  File "/app/lib/python3.5/site-packages/mutagen/_util.py", line 140, in wrapper
    return func(self, h, *args, **kwargs)
  File "/app/lib/python3.5/site-packages/mutagen/id3/_file.py", line 255, in save
    f.write(data)
OSError: [Errno 38] Function not implemented

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/lib/python3.5/site-packages/picard/util/thread.py", line 47, in run
    result = self.func()
  File "/app/lib/python3.5/site-packages/picard/file.py", line 211, in _save_and_rename
    self._save(old_filename, metadata)
  File "/app/lib/python3.5/site-packages/picard/formats/id3.py", line 418, in _save
    self._save_tags(tags, encode_filename(filename))
  File "/app/lib/python3.5/site-packages/picard/formats/id3.py", line 525, in _save_tags
    tags.save(filename, v2_version=3, v1=v1, v23_sep=separator)
  File "/app/lib/python3.5/site-packages/mutagen/_util.py", line 173, in wrapper
    reraise(exc_dest, err, sys.exc_info()[2])
  File "/app/lib/python3.5/site-packages/mutagen/_compat.py", line 86, in reraise
    raise tp(value).with_traceback(tb)
  File "/app/lib/python3.5/site-packages/mutagen/_util.py", line 169, in wrapper
    return func(*args, **kwargs)
  File "/app/lib/python3.5/site-packages/mutagen/_util.py", line 140, in wrapper
    return func(self, h, *args, **kwargs)
  File "/app/lib/python3.5/site-packages/mutagen/id3/_file.py", line 255, in save
    f.write(data)
mutagen.id3._util.error: [Errno 38] Function not implemented
```

I'm not sure if this is related to Flatpak or Picard. Ping @TingPing @bochecha. WDYT ?

**Comment by bochecha** [comment]
@Lctrs sorry, no idea.

I'll see if I can get some time to try and reproduce it and figure it out.

**Comment by Lctrs** [comment]
@bochecha I tried several things to get rid of this error without success : 
- Building with latest Python version (3.6.5)
- Building with GCC 7
- Both

**Comment by bochecha** [comment]
I'll see if I can get some time to try and reproduce it and figure it out.I just tried building your branch exactly as it is.

It worked just fine for me.

How did you test it?

Here's what I did:

```
$ flatpak-builder --disable-cache --force-clean --ccache --repo=repo app org.musicbrainz.Picard.json
$ flatpak remote-add --user --no-gpg-verify local-picard2 repo
$ flatpak install --user local-picard2 org.musicbrainz.Picard
```

Then I ran Picard from GNOME Shell, added some music files, modified their tags and saved them.

**Comment by bochecha** [comment]
I couldn't reproduce your issue with a file in a folder containing non-ascii characters.

I think you're hitting a variation of #1 which was already reported with Picard 1. Let's try and investigate the issue there, and focus this pull request on Picard 2.

**Comment by Lctrs** [comment]
@bochecha Picard 2.0 has been tagged in [Picard's github repo](https://github.com/metabrainz/picard/releases).

This PR can now be merged.

**Comment by Lctrs** [comment]
Also officially announced on the [Musicbrainz's blog](https://blog.musicbrainz.org/2018/07/19/picard-2-0-released/).

**Comment by Lctrs** [comment]
bot, build

**Comment by Lctrs** [comment]
Test build is here : https://flathub.org/builds/#/builders/2/builds/2956.

**Comment by bochecha** [comment]
Hi, thanks for following through with this.

I've been a bit busy lately, but I'll try to get to this ASAP.

**Comment by nedrichards** [comment]
bot, build

**Comment by bochecha** [comment]
Alright, so I had a few little changes I wanted from this branch, but then I've already made you wait so long that it didn't seem right to just now make you change such small things.

I added the changes myself, and merged your branch manually. It's in master now, and should be available in the repo soon.

Thanks a lot for this contribution, thanks for your patience, and sorry I've been so slow to reply. (freedesktop-sdk kept me busy for some time :smile:)

**Comment by Lctrs** [comment]
I haven't made the switch to Qt and PyQt 5.11 because it's not officially supported by upstream. See their [requirements.txt](https://github.com/metabrainz/picard/blob/release-2.0.2/requirements.txt) file for the current release.

Even though I had no problem when testing the app locally with PyQt 5.11.

**Comment by bochecha** [comment]
> See their requirements.txt file for the current release.

Oh right… Oops! :confounded: 

I'll fix, thanks for finding this.
